tiledb/sm/array/array.cc

```diff
@@ -811,61 +830,105 @@ Array::get_enumerations_all_schemas() {
+              ") has no loaded schema named '" + schema_enmrs.first +
+              "'; If the array was recently evolved be sure to reopen it after "
+              "applying the evolution.");
         }
-        enmrs_to_load.insert(enmr_name);
-      }
 
-      // Create a vector of paths to be loaded.
-      std::vector<std::string> paths_to_load;
-      for (auto& enmr_name : enmrs_to_load) {
-        auto path = schema.second->get_enumeration_path_name(enmr_name);
-        paths_to_load.push_back(path);
+        auto schema = array_schemas_all().at(schema_enmrs.first);
+        for (const auto& enmr : schema_enmrs.second) {
+          if (!schema->is_enumeration_loaded(enmr->name())) {
+            schema->store_enumeration(enmr);
+          }
+          // Also store enumerations into the latest schema when we encounter
+          // it.
+          if (schema_enmrs.first == latest_schema->name()) {
+            if (!latest_schema->is_enumeration_loaded(enmr->name())) {
+              latest_schema->store_enumeration(enmr);
+            }
+          }
+        }
       }
+    } else {
+      auto latest_schema = opened_array_->array_schema_latest_ptr();
+      for (const auto& schema : array_schemas_all()) {
+        std::unordered_set<std::string> enmrs_to_load;
+        auto enumeration_names = schema.second->get_enumeration_names();
+        // Dedupe requested names and filter out anything already loaded.
+        for (auto& enmr_name : enumeration_names) {
+          if (schema.second->is_enumeration_loaded(enmr_name)) {
+            continue;
+          }
+          enmrs_to_load.insert(enmr_name);
+        }
 
-      // Load the enumerations from storage
-      auto loaded = array_directory().load_enumerations_from_paths(
-          paths_to_load, *encryption_key(), memory_tracker_);
+        // Create a vector of paths to be loaded.
+        std::vector<std::string> paths_to_load;
+        for (auto& enmr_name : enmrs_to_load) {
+          auto path = schema.second->get_enumeration_path_name(enmr_name);
+          paths_to_load.push_back(path);
+        }
 
-      // Store the loaded enumerations in the schema.
-      for (auto& enmr : loaded) {
-        schema.second->store_enumeration(enmr);
+        // Load the enumerations from storage
+        auto loaded = array_directory().load_enumerations_from_paths(
+            paths_to_load, *encryption_key(), memory_tracker_);
+
+        // Store the loaded enumerations in the schema.
+        for (auto& enmr : loaded) {
+          if (!schema.second->is_enumeration_loaded(enmr->name())) {
+            schema.second->store_enumeration(enmr);
+          }
+          // Also store enumerations into latest schema.
+          if (schema.first == latest_schema->name()) {
+            if (!latest_schema->is_enumeration_loaded(enmr->name())) {
+              latest_schema->store_enumeration(enmr);
+            }
+          }
+        }
+        ret[schema.first] = loaded;
```

Comment:
Doesn't this mean that what you return will only contain the enumerations you just loaded from disk and not all of them (loaded from disk + already loaded)?
Should we add som more test coverage for the return value of this function?

Reply:
Nice catch, it was surprisingly easy to fix but difficult to get testing setup. I added two new tests that check the return value, one for loading all enums and one for partially loading enumerations where some were already loaded before we call `get_enumerations_all_schemas`. If you're not a fan of the [equality operators](https://github.com/TileDB-Inc/TileDB/pull/5360/commits/4520d0aeb31030cc55d0201186699e346ca4e5dc#diff-b43039cae2853cf7c0c022a4eee6023956479b58a906f18188f898da930ccec9R73) I could write a test helper method instead.

Reply:
Thanks for taking the time to implement those tests although cumbersome. I think Ryan needed to implement a comparator for enumerations too [here](https://github.com/TileDB-Inc/TileDB/pull/5359/files#diff-652a3de9a9aa89aa0e8fb99e701c60b074250f84d55c551d8052d3b665cfc22bR42), in case you'd prefer to somehow reuse that one (enumerations are trending right now in our PRs :stuck_out_tongue:). I don't mind really as long as we have tests that work.